src/main/java/com/radixdlt/atomos/ParticleDefinition.java

```diff
@@ -58,7 +63,88 @@
 		return rriMapper;
 	}
 
+	Function<T, Spin> getVirtualizeSpin() {
+		return virtualizeSpin;
+	}
+
 	public boolean allowsTransitionsFromOutsideScrypts() {
 		return allowsTransitionsFromOutsideScrypts;
 	}
+
+	/**
+	 * Creates a Builder for {@link ParticleDefinition}s.
+	 * @param <T> The type of the {@link ParticleDefinition}
+	 * @return The builder
+	 */
+	public static <T extends Particle> Builder<T> builder() {
+		return new Builder<>();
+	}
+
+	/**
+	 * A Builder for {@link ParticleDefinition}s
+	 * @param <T> The type of the particle to be defined
+	 */
+	public static class Builder<T extends Particle> {
+		private Function<T, Set<RadixAddress>> addressMapper;
+		private Function<T, Result> staticValidation = x -> Result.success();
+		private Function<T, RRI> rriMapper;
+		private Function<T, Spin> virtualizeSpin;
+		private boolean allowsTransitionsFromOutsideScrypts;
+
+		private Builder() {
+		}
+
+		public Builder<T> singleAddressMapper(Function<T, RadixAddress> addressMapper) {
+			this.addressMapper = p -> Collections.singleton(addressMapper.apply(p));
+			return this;
+		}
+
+		public Builder<T> addressMapper(Function<T, Set<RadixAddress>> addressMapper) {
+			this.addressMapper = addressMapper;
+			return this;
+		}
+
+		public Builder<T> staticValidation(Function<T, Result> staticValidation) {
+			this.staticValidation = staticValidation;
+			return this;
+		}
+
+		public Builder<T> rriMapper(Function<T, RRI> rriMapper) {
+			this.rriMapper = rriMapper;
+			return this;
+		}
+
+		public Builder<T> virtualizeSpin(Function<T, Spin> virtualizeSpin) {
+			this.virtualizeSpin = virtualizeSpin;
+			return this;
+		}
+
+		public Builder<T> allowTransitionsFromOutsideScrypts() {
+			this.allowsTransitionsFromOutsideScrypts = true;
+			return this;
+		}
+
+		/**
+		 * Builds the {@link ParticleDefinition} with the given properties, casting as necessary.
+		 * All properties except the address mapper are optional.
+		 *
+		 * TODO Cleanup: Recasting the Particle type here is non-ideal, but fixing it would take too long.
+		 * @param <U> The type of the built ParticleDefinition (must be castable from builder type T)
+		 * @return The built {@link ParticleDefinition}
+		 */
+		public <U extends Particle> ParticleDefinition<U> build() {
+			if (addressMapper == null) {
+				throw new IllegalStateException("addressMapper is required");
+			}
+
+			// cast as necessary
+			return new ParticleDefinition<>(
```

Comment:
This is a mess from a type safety point of view, and I appreciate that it's not a mess that you made entirely by yourself.  I really think this should be tidied up.

Reply:
Totally agreed. I've added comment here explaining as much. Unfortunately, fixing this properly would require some more substantial changes, so it will have to do for now.